src/server.c

```diff
@@ -241,6 +246,33 @@ static bool handle_server_startup(PgSocket *server, PktHdr *pkt)
 	return res;
 }
 
+/*
+ * Compare connection attributes as returned by the server
+ * against desired connection attributes. This matches the
+ * behavior of libpq's `target_session_attrs` albeit only
+ * implementing a subset of the functionality:
+ * 1) It does not support `prefer-standby`.
+ * 2) It requires PG 14+ to send connection parameters at startup.
+ */
+bool valid_target_session_attrs(PgSocket *server)
+{
+	VarCache *v = &server->vars;
+	const struct PStr *in_hot_standby = v->var_list[VInHotStandby];
+	const struct PStr *default_transaction_read_only = v->var_list[VDefaultTransactionReadOnly];
+	enum TargetSessionAttrs target = server->pool->db->target_session_attrs;
+
+	// If the server did not return the in_hot_standby and/or
+	// default_transaction_read_only parameters, assume false.
+	bool hot_standby = in_hot_standby && strcmp(in_hot_standby->str, "on") == 0;
+	bool transaction_read_only = default_transaction_read_only && strcmp(default_transaction_read_only->str, "on") == 0;
+
+	return (target == TARGET_SESSION_ANY) ||
+	       (target == TARGET_SESSION_READWRITE && !transaction_read_only) ||
+	       (target == TARGET_SESSION_READONLY && transaction_read_only) ||
```

Comment:
These are wrong. The correct logic is as follows:
```suggestion
	       (target == TARGET_SESSION_READWRITE && !transaction_read_only && !hot_standby) ||
	       (target == TARGET_SESSION_READONLY && (transaction_read_only || hot_standby)) ||
```

I'd prefer to create separate variable for that though for readability. Something like the following and then use that:
```
bool read_only = default_transaction_read_only || hot_standby;
```

Reply:
This is indeed the correct logic but am having difficulty exposing this in a test. I think because: in PG14+ PostgreSQL servers begin to put `transaction_read_only` into the session and it is always set to `on` when in hot standby. Is there a scenario I can add coverage for where `transaction_read_only = off` and `in_hot_standby = true`?

Thanks for reasoning through the logic miss and the suggested clean-up 👍

Reply:
8af5a6fdc5c3d7832c8b2fd7ad16c51caf4b0870

Reply:
Hmm, that does make testing one variant impossible I guess. But you can still test that read-only accepts both a hot_standby, and a database specifially marked as read only by default using `ALTER DATABASE ... SET transaction_read_only = true` (without putting it in hot_standby mode).

Reply:
psycopg makes this annoying. I might need to use psql to `ALTER DATABASE` as `Postgres#configure` is annoyingly giving `psycopg.errors.CantChangeRuntimeParam: parameter "transaction_read_only" cannot be changed`

Reply:
Ah silly me `configure` is `ALTER SYSTEM`. Not psycopg's fault at all.

Reply:
I believe this feedback is addressed in f752642b97781680f07fd8232deda7dcef16cae0 and 4ef999a6a2d7a0cc99d5449698b875c1284a2203 please unresolve if you disagree.